src/IFrameNavigator.ts

```diff
@@ -876,16 +804,115 @@ export default class IFrameNavigator implements Navigator {
       return manifest;
     } catch (err) {
       this.abortOnError(err);
-      // abortOnError should already throw the Error, this is here to make TS happy.  
-      // We should try to remove the giant try/catch block to avoid this.  
-      throw new Error(err);
+      throw new Error("Could not load manifest " + err);
+    }
+  }
+
+  private async createTOC(manifest: Manifest): Promise<void> {
+    const createTOC = (parentElement: Element, links: Array<Link>) => {
+      const listElement: HTMLOListElement = document.createElement("ol");
+      let lastLink: HTMLAnchorElement | null = null;
+      for (const link of links) {
+        const listItemElement: HTMLLIElement = document.createElement("li");
+        const linkElement: HTMLAnchorElement = document.createElement("a");
+        const spanElement: HTMLSpanElement = document.createElement("span");
+        linkElement.tabIndex = -1;
+        let href = "";
+        if (link.href) {
+          href = new URL(link.href, this.manifestUrl.href).href;
+
+          linkElement.href = href;
+          linkElement.innerHTML = link.title || "";
+          listItemElement.appendChild(linkElement);
+        } else {
+          spanElement.innerHTML = link.title || "";
+          listItemElement.appendChild(spanElement);
+        }
+        if (link.children && link.children.length > 0) {
+          createTOC(listItemElement, link.children);
+        }
+
+        listElement.appendChild(listItemElement);
+        lastLink = linkElement;
+      }
+
+      // Trap keyboard focus inside the TOC while it's open.
+      if (lastLink) {
+        this.setupModalFocusTrap(this.tocView, this.contentsControl, lastLink);
+      }
+
+      listElement.addEventListener("click", (event: Event) => {
+        event.preventDefault();
+        event.stopPropagation();
+        if (
+          event.target &&
+          (event.target as HTMLElement).tagName.toLowerCase() === "a"
+        ) {
+          let linkElement = event.target as HTMLAnchorElement;
+
+          if (linkElement.className.indexOf("active") !== -1) {
+            // This TOC item is already loaded. Hide the TOC
+            // but don't navigate.
+            this.hideTOC();
+          } else {
+            // Set focus back to the contents toggle button so screen readers
+            // don't get stuck on a hidden link.
+            this.contentsControl.focus();
+            this.navigate({
+              resource: linkElement.href,
+              position: 0,
+            });
+          }
+        }
+      });
+
+      parentElement.appendChild(listElement);
+    };
+
+    if (manifest.toc.length) {
```

Comment:
This might be source of the `cannot read length of undefined` error. Just a guess